Add Feature Model Cloning

s9latimm opened the pull request:
- Adds deep cloning for `FeatureModel` and related data structures (`Constraint`, `FeatureSourceRange`). 
- Reworks `getParentFeature` replacing `getParent` with a template in `FeatureTreeNode`.

s9latimm committed:
Working clone and tests

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureSourceRange.h
M	include/vara/Feature/FeatureTreeNode.h
M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModel.cpp

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -113,6 +115,12 @@ class OrConstraint : public BinaryConstraint, public BooleanConstraint {
 
+  std::unique_ptr<Constraint> clone() override {
+    return std::make_unique<OrConstraint>(
+        std::move(this->getLeftOperand()->clone()),
+        std::move(this->getRightOperand()->clone()));
```
```suggestion
        this->getLeftOperand()->clone(),
        this->getRightOperand()->clone());
```
The returned value is already an rvalue

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -130,6 +138,12 @@ class XorConstraint : public BinaryConstraint, public BooleanConstraint {
 
+  std::unique_ptr<Constraint> clone() override {
+    return std::make_unique<XorConstraint>(
+        std::move(this->getLeftOperand()->clone()),
+        std::move(this->getRightOperand()->clone()));
```
```suggestion
        this->getLeftOperand()->clone(),
        this->getRightOperand()->clone());
```

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -147,6 +161,12 @@ class AndConstraint : public BinaryConstraint, public BooleanConstraint {
                          std::move(RightOperand)) {}
 
+  std::unique_ptr<Constraint> clone() override {
+    return std::make_unique<AndConstraint>(
```
etc...

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -481,10 +587,14 @@ class Feature;
   [[nodiscard]] Feature *getFeature() const;
 
+  std::unique_ptr<Constraint> clone() override {
+    return std::make_unique<PrimaryFeatureConstraint>(this->getFeature());
```
Is this feature ptr later replaced?

s9latimm replied:
I reworked this part, so that this copy only holds a new shallow feature (name only), which later needs replacement (the `FeatureModelBuilder` used cloning a `FeatureModel` already does exactly that).

Also, added comment should explain the intended behavior.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -50,18 +50,6 @@ class Feature : public FeatureTreeNode {
 
-  /// Search parent feature in tree structure -- this may not exist or nullptr
-  /// if node is already root.
-  [[nodiscard]] Feature *getParentFeature() const {
```
Why did you move this to the `TreeNodeClass`? As an abstraction this makes more sense on a Feature, with the move the TreeNode is now hard Tied to the Feature.

s9latimm replied:
`Relationship` now needs a look up to a parent feature too. I reworked this as a template in `FeatureTreeNode`.

vulder replied:
I fixed the Relationship with local cast handling, we don't really need the template.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -225,12 +213,13 @@ class BinaryFeature : public Feature {
       string Name, bool Opt = false,
       std::vector<FeatureSourceRange> Loc = std::vector<FeatureSourceRange>(),
-      Feature *Parent = nullptr, const NodeSetType &Children = {})
+      FeatureTreeNode *Parent = nullptr, const NodeSetType &Children = {})
```
A gain, more abstraction breaks. I don't think we should rework this.

vulder reviewed (changes requested):
Hmm, I think we should not break the Feature/FeatureTreeNode abstraction. Why did you mix these two to implement clone?

s9latimm committed:
Requested changes and some improvements

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureTreeNode.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/FeatureModel.cpp

s9latimm committed:
Rework constraint cloning

M	include/vara/Feature/Constraint.h
M	lib/Feature/Constraint.cpp

s9latimm committed:
Inline

M	include/vara/Feature/FeatureTreeNode.h

vulder committed:
Fix relationship without cast template

vulder commented on unittests/Feature/FeatureModel.cpp:
```diff
@@ -13,6 +13,79 @@ TEST(FeatureModel, build) {
+  FM.reset();
+
+  // NOLINTNEXTLINE
+  ASSERT_EXIT(EXPECT_TRUE(FM->getFeature("a")->isRoot()),
+              testing::KilledBySignal(SIGSEGV), ".*");
```
Why do you need the EXPECT_TRUE inside the EXIT? Should the segfault not also occur without it?

vulder committed:
Small fixes

s9latimm committed:
Fix bindings

M	bindings/python/vara-feature/pybind_Feature.cpp

s9latimm committed:
Merge branch 'f-FeatureModelClone' of github.com:se-sic/vara-feature into f-FeatureModelClone

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureTreeNode.h
M	lib/Feature/Constraint.cpp
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/FeatureModel.cpp

s9latimm committed:
Add comment

M	unittests/Feature/FeatureModel.cpp

s9latimm committed:
Fix bindigs -- this time for real

M	bindings/python/vara-feature/pybind_Feature.cpp
